lib/spack/spack/directives.py

```diff
@@ -598,6 +600,23 @@ def _execute_depends_on(pkg: "spack.package_base.PackageBase"):
     return _execute_depends_on
 
 
+@directive(dicts=())
+def redistribute(source=True, binary=True):
```

Comment:
Ok sorry I should've thought about this on the first review, but I am *positive* that people are going to want to do this conditionally (another good reason for a directive).  There are packages where, depending on the version, or some variants, the package may be redistributable or not.  Also note that `license()` supports `when=`, and we'll want to set this depending on the license.

So:

```suggestion
def redistribute(source: bool = True, binary: bool = True, *, when: WhenType = None):
```